airflow/contrib/operators/gcp_function_operator.py

```diff
@@ -0,0 +1,571 @@
+
+
+class SourceCodeValidation:
+    """
+    Validates Source Code parameters.
+
+    Responsible for validating if we have enough information about the source code of
+    the GCF function. The source code to be used can be specified using different options:
+    * non-empty sourceUploadUrl in case the code was just uploaded using
+      "generateUploadUrl" mechanism
+    * empty sourceUploadUrl and path to zip file (then generateUploadUrl will
+      be used to upload the zip file and sourceUploadUrl will be set appropriately
+    * archiveUrl that points to valid Google Storage link (starting with gs://)
+      where the zip file is uploaded.
+
+    :param body: Body passed to the create/update method calls.
+    :type body: dict
+    :param zip_path: path to the zip file containing source code. If it is
+        set, then sourceUploadUrl should not be specified in the body (or it should
+        be empty), then the zip file will be uploaded using upload URL generated
+        via generateUploadUrl from cloud functions API
+    :type body: str
+
+    """
+    upload_function = None
+
+    def __init__(self, body, zip_path):
+        self.body = body
+        self.zip_path = zip_path
+
+    @staticmethod
+    def _is_present_and_empty(dictionary, field):
+        return field in dictionary and not dictionary[field]
+
+    def _is_upload_url_and_no_zip_path(self):
+        if self._is_present_and_empty(self.body, SOURCE_UPLOAD_URL):
+            if not self.zip_path:
+                raise AirflowException(
+                    "If parameter '{}' is empty in 'body' then argument '{}' needed.".
+                    format(SOURCE_UPLOAD_URL, ZIP_PATH))
+
+    def _is_upload_url_and_zip_path(self):
+        if SOURCE_UPLOAD_URL in self.body and self.zip_path:
+            if not self.body[SOURCE_UPLOAD_URL]:
+                self.upload_function = True
+            else:
+                raise AirflowException("Only one of '{}' in body or '{}' argument "
+                                       "possible. Found both."
+                                       .format(SOURCE_UPLOAD_URL, ZIP_PATH))
+
+    def _is_archive_url_and_zip_path(self):
+        if SOURCE_ARCHIVE_URL in self.body and self.zip_path:
+            raise AirflowException("Only one of '{}' in body or '{}' argument "
+                                   "possible. Found both."
+                                   .format(SOURCE_ARCHIVE_URL, ZIP_PATH))
+
+    def should_upload_function(self):
+        if self.upload_function is None:
+            raise AirflowException('validate() method has to be invoked first')
+        return self.upload_function
+
+    def validate(self):
+        self._is_archive_url_and_zip_path()
+        self._is_upload_url_and_zip_path()
+        self._is_upload_url_and_no_zip_path()
+        if self.upload_function is None:
+            self.upload_function = False
+
+
+FUNCTION_NAME_PATTERN = '^projects/[^/]+/locations/[^/]+/functions/[^/]+$'
```

Comment:
This is a nit pick, and feel free to ignore, but consider making these constants compiled expressions (i.e. move 'compile' to here). IIUC, it will mean the regexps will be compiled at startup, so any syntax errors would be called earlier.

Reply:
Why not. It's cheap to move it :)